client/widmgr/widmgr.go

```diff
@@ -4,80 +4,367 @@
+func (m *WIDMgr) Watch(id cstructs.TaskIdentity) (<-chan *structs.SignedWorkloadIdentity, func()) {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
+
+	// If Shutdown has been called return a closed chan
+	if m.stopCtx.Err() != nil {
+		c := make(chan *structs.SignedWorkloadIdentity)
+		close(c)
+		return c, func() {}
+	}
+
+	//TODO I think we only want 1 watcher per id, so let's just close on double watches?
+	if existing, ok := m.watchers[id]; ok {
+		close(existing)
+	}
+
+	// Buffer of 1 so sends don't block on receives
+	c := make(chan *structs.SignedWorkloadIdentity, 1)
+	m.watchers[id] = c
+
+	// Create a cancel func for watchers to deregister when they exit.
+	cancel := func() {
+		m.watchersLock.Lock()
+		defer m.watchersLock.Unlock()
+
+		delete(m.watchers, id)
+
+		//TODO should we close(c) here? seems sketchy... caller shouldn't try to
+		//recv on c after calling cancel
+	}
+
+	// Prime chan with latest token to avoid a race condition where consumers
+	// could miss a token update between Get and Watch calls.
+	if token := m.get(id); token != nil {
+		c <- token
+	}
+
+	return c, cancel
+}
+
+// Shutdown stops renewal and closes all watch chans.
+func (m *WIDMgr) Shutdown() {
+	m.watchersLock.Lock()
+	defer m.watchersLock.Unlock()
+
+	m.stop()
+
+	for _, c := range m.watchers {
+		close(c)
+	}
+}
+
+// getIdentities fetches all signed identities or returns an error.
+func (m *WIDMgr) getIdentities() error {
+	m.lastTokenLock.Lock()
+	defer m.lastTokenLock.Unlock()
+
+	if len(m.widSpecs) == 0 {
+		return nil
+	}
+
+	reqs := make([]*structs.WorkloadIdentityRequest, 0, len(m.widSpecs))
+	for taskName, widspecs := range m.widSpecs {
+		for _, widspec := range widspecs {
+			reqs = append(reqs, &structs.WorkloadIdentityRequest{
+				AllocID:      m.allocID,
+				TaskName:     taskName,
+				IdentityName: widspec.Name,
+			})
+		}
+	}
+
+	// Get signed workload identities
+	signedWIDs, err := m.signer.SignIdentities(m.minIndex, reqs)
+	if err != nil {
+		return err
+	}
+
+	// Index initial workload identities by name
+	m.lastToken = make(map[cstructs.TaskIdentity]*structs.SignedWorkloadIdentity, len(signedWIDs))
+	for _, swid := range signedWIDs {
+		id := cstructs.TaskIdentity{
+			TaskName:     swid.TaskName,
+			IdentityName: swid.IdentityName,
+		}
+
+		m.lastToken[id] = swid
+	}
+
+	return nil
+}
+
+// renew fetches new signed workload identity tokens before the existing tokens
+// expire.
+func (m *WIDMgr) renew() {
+	if len(m.widSpecs) == 0 {
+		return
+	}
+
+	reqs := make([]*structs.WorkloadIdentityRequest, 0, len(m.widSpecs))
+	for taskName, widspecs := range m.widSpecs {
+		for _, widspec := range widspecs {
+			if widspec.TTL == 0 {
+				continue
+			}
+			reqs = append(reqs, &structs.WorkloadIdentityRequest{
+				AllocID:      m.allocID,
+				TaskName:     taskName,
+				IdentityName: widspec.Name,
+			})
+		}
+	}
+
+	renewNow := false
+	minExp := time.Now().Add(30 * time.Hour) // set high default expiration
```

Comment:
hmm, but what would the initial `minExp` value be then? We only know the `token.Expiration` once we look at the tokens themselves, not sure how would we get the `minExp.IsZero()` to work in this case.

Reply:
> hmm, but what would the initial minExp value be then? 

Zero. `minExp.IsZero()` would mean "this is the first iteration, so set the value without comparing it". It also handles the case where TTL is 0 as `minExp` will still be zero and set to the first non-zero TTL.

Looking at the code again, it could also serve as a signal that all identities have `ttl = 0` which means we could even exit the `renew` function? 🤔 

I also noticed later that this logic is being used below as well:
https://github.com/hashicorp/nomad/pull/18431/files/265f6cdc29f7424039fa749b80ca320662151c92#r1330575709

Reply:
I like the idea of being able to bail out of the renew loop entirely. That implies downstream consumers will need to _either_ `Get` or `Watch` based on whether they have a TTL, but I think that's ok.

Reply:
I addressed the `minExp` value in eb1aba34f92cb4fc5cc9bc8c1bc60788b81397d8, and for now we decided to keep the `Get` method.